# Chapter 3 Practice Lab

In [1]:
import pandas as pd
import os.path as osp
import statsmodels.api as sm
import statsmodels.tools as smtools
import statsmodels.stats as smstats
import numpy as np
import scipy.stats as stats
import seaborn as sns
datasets_path = osp.join('../datasets')

8 - Perform a simple linear regression with
mpg as the response and horsepower as the predictor.
*  Is there a relationship between the predictor and the response?
*  How strong is the relationship between the predictor and
the response?
*  How strong is the relationship between the predictor and
the response?
*  What is the predicted mpg associated with a horsepower of
98? What are the associated 95% confidence and prediction
intervals?


Plot the response and the predictor. Use the abline() function
to display the least squares regression line.
Use the plot() function to produce diagnostic plots of the least
squares regression fit. Comment on any problems you see with
the fit.

In [29]:
auto = pd.read_csv(osp.join(datasets_path, 'auto.csv'))

In [6]:
auto.head(3)

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite

In [30]:
auto = auto.loc[auto['horsepower'] != '?', :]

In [32]:
predictor = smtools.tools.add_constant(auto['horsepower'].astype('float'))
target = auto['mpg']
simple_lm = sm.OLS(target, predictor)
simple_lm_fit = simple_lm.fit()

In [35]:
simple_lm_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.606
Model:                            OLS   Adj. R-squared:                  0.605
Method:                 Least Squares   F-statistic:                     599.7
Date:                Thu, 21 May 2020   Prob (F-statistic):           7.03e-81
Time:                        12:33:40   Log-Likelihood:                -1178.7
No. Observations:                 392   AIC:                             2361.
Df Residuals:                     390   BIC:                             2369.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         39.9359      0.717     55.660      0.000      38.525      41.347
horsepower    -0.1578      0.006    -24.489      0.000      -0.171      -0.145
==============================================================================
Omnibus:                       16.432   Durbin-Watson:                   0.920
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               17.305
Skew:                           0.492   Prob(JB):                     0.000175
Kurtosis:                       3.299   Cond. No.                         322.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Is there a relationship between the predictor and the response?**


Yes, there is a relationship between horsepower and mpg. This is based on the low p-value we get for the horsepower predictor, which means that the probability of getting the value -0.1578 given that the true value is 0 is super low. 


In [55]:
RSE = np.sqrt(simple_lm_fit.scale)
RSE, target.mean(), RSE * 100 / target.mean() 

(4.90575691954594, 23.445918367346938, 20.923714066914833)

**How strong is the relationship between the predictor and
the response?**

The $R^2$ metric is ca. 0.6, which indicates that the fit explains 60% of the variance of the response variable which in turn indicates that there must be a relationship between the only predictor and the independent variable. 

The $RSE$ (Residual Standard Error), is ca. 5 so wrt the mean of the response variable it indicates a mean percentage of error of 20%. So there is a considerable relationshipt between the predictor and response 

**Is the relationship between the predictor and the response
positive or negative?**

It is negative, because the estimated coefficient is -0.1578, this means that if we increase our car horspower we will decrease our mile per gallon performance.

**What is the predicted mpg associated with a horsepower of
98? What are the associated 95% confidence and prediction
intervals?**

In [75]:
simple_lm_fit.get_prediction(np.array([1.0, 98.0])).summary_frame()

mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  24.467077  0.251262      23.973079      24.961075     14.809396   

   obs_ci_upper  
0     34.124758

In [78]:
predicted_value = simple_lm_fit.predict(np.array([1.0, 98.0]))
p_interval = (predicted_value-1.96*RSE,predicted_value+1.96*RSE)
p_interval

(array([14.85179359]), array([34.08236071]))

The predicted mpg is 24.467077, the Confidence Interval is equal to $[14.809396, 34.124758]$, while the prediction interval is 

In [79]:
pd.read_csv(osp.join(datasets_path, 'generacion-de-residuos-solidos-urbanos-por-provincia.csv'))

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe1 in position 12: invalid continuation byte